netty/src/main/java/io/grpc/netty/NettyServer.java

```diff
@@ -135,7 +138,31 @@ public void operationComplete(ChannelFuture future) throws Exception {
           log.log(Level.WARNING, "Error shutting down server", future.cause());
         }
         listener.serverShutdown();
+        releaseSharedGroups();
       }
     });
   }
+
+  private void allocateSharedGroups() {
+    if (bossGroup == null) {
+      bossGroup = SharedResourceHolder.get(Utils.DEFAULT_BOSS_EVENT_LOOP_GROUP);
+    }
+    if (workerGroup == null) {
+      workerGroup = SharedResourceHolder.get(Utils.DEFAULT_WORKER_EVENT_LOOP_GROUP);
+    }
+  }
+
+  private void releaseSharedGroups() {
+    try {
+      if (usingSharedBossGroup && bossGroup != null) {
+        SharedResourceHolder.release(Utils.DEFAULT_BOSS_EVENT_LOOP_GROUP, bossGroup);
+      }
+      if (usingSharedWorkerGroup && workerGroup != null) {
+        SharedResourceHolder.release(Utils.DEFAULT_WORKER_EVENT_LOOP_GROUP, workerGroup);
+      }
+    } finally {
+      bossGroup = null;
+      workerGroup = null;
```

Comment:
:-/ If we are concerned about exceptions, then if releasing the boss throws an exception the worker group won't be released. Either don't deal with the exceptions, or make the exception handling correct.